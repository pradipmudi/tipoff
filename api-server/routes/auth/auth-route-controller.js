import express from 'express';
import log4js from 'log4js';
import { check, validationResult, body } from 'express-validator';

import RouteUtils from '../../utils/route-utils';
import UserSvc from '../../services/users/user-svc';

const log = log4js.getLogger("auth-route-ctrl");
const userSvc = new UserSvc();
const router = express.Router();

/* POST Login API.  
/api/v1/auth/login
{
    email: 'required',
    password: 'required'
}
*/
router.post('/login',
	[
		//validate input
        check('userId',RouteUtils.INVALID_EMAIL_ERR_MSG).isEmail(),
		check('password',RouteUtils.INVALID_PASSWORD).isLength({min: 6, max: 64})
	], 
	async (req, res) => {
		RouteUtils.logFunctionEntryInDebug(log, `in POST /login ` , req.id);
		//Check if validation resulted in Errors
		const errors = validationResult(req);
		if (!errors.isEmpty()) {
			//return 422 - Unprocessable Entity with errors
			return RouteUtils.sendUnprocessableEntity(res,errors,log, req.id);
		}
		try{
			let resp = await userSvc.validateUser(req)
				.catch(error => { throw error; });
			if(resp && resp.validUser){
				req.loggedInUser = req.body.userId;
				return RouteUtils.sendSuccessResponse(res,resp,undefined, log, 
					`returning successful login `,req.id);
			}else{
				return RouteUtils.sendAuthorizationFailure(res,`login unsuccessful - ${req.body.userId}`,log,req.id)
			}
			
		} catch(error) {
			return RouteUtils.sendFailureResponse(res, error, undefined, log, req.id);
		}
	}
);

router.post('/register',
	[
		//validate input
		check('userId',RouteUtils.INVALID_EMAIL_ERR_MSG).isEmail(),
		check('name',RouteUtils.INVALID_NAME).isLength({min:3}),
        check('password',RouteUtils.INVALID_PASSWORD).isLength({min: 6, max: 64}),
		check('confirmPassword', RouteUtils.INVALID_CONFIRM_PWD)
			.custom((value, { req }) => value === req.body.password)
        
	], 
	async (req, res) => {
		RouteUtils.logFunctionEntryInDebug(log, `in PUT /register ` , req.id);
		//Check if validation resulted in Errors
		const errors = validationResult(req);
		if (!errors.isEmpty()) {
			//return 422 - Unprocessable Entity with errors
			return RouteUtils.sendUnprocessableEntity(res,errors,log, req.id);
		}
		try {	
			const resp = await userSvc.addUser(req)
				.catch(error => { throw error; });
			return RouteUtils.sendSuccessResponse(res,resp, 201,  log, 
				`returning successful user `,req.id);
		} catch(error) {
			return RouteUtils.sendFailureResponse(res, error, undefined, log, req.id);
		}
	}
);

export default router;
