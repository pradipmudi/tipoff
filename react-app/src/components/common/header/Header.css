.logo {
    font-size: 2em;
    font-weight: 900;
    color:#e1f4f3 !important;
    margin-left:1em;
    cursor:pointer;
}

.logo-text {
    font-size: 1.8em;
    font-weight: 700;
    color:#e1f4f3 !important;
    cursor:pointer;
}

.user-name{
    color:#e1f4f3 !important;
}

.header {
    background-color:#333333 !important;
    color:#e1f4f3 !important;
}

.user-name .fas {
    color:#e1f4f3 !important;
}

.navbar-burger{
    background-color: #333;
}

.navbar-burger span{
    color: #fff;
}