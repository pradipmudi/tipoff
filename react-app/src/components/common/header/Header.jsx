import React, {useState} from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { 
    Navbar, 
    NavbarStart,
    NavbarBrand, 
    NavbarBurger, 
    NavbarMenu, 
    NavbarItem, 
    Icon,
    NavbarEnd,
    Field,
    Control,
    Button
 } from 'bloomer';

import auth from '../../../service/auth-svc';
import {USER_SIGN_OUT} from '../../../redux/user/user-constants';
import {selectCurrentUser} from '../../../redux/user/user-selectors';

import './Header.css';
const Header = () => {

    const history = useHistory();
    const currentUser = useSelector(selectCurrentUser, shallowEqual);
    const dispatch = useDispatch();

    const [state,setState] = useState({ 
        isActive: false, 
        isDropdownOpen: false
    });

    const onClickNav = (e) => {
        e.preventDefault();
        setState({
            ...state,
            isActive: !state.isActive, 
            isDropdownOpen: false
        });
    }

    const signOut = (e) => {
        e.preventDefault();
        dispatch({type:USER_SIGN_OUT})
        auth.logout(() => {
            history.push("/");
        });
    }

    const navigateToHome = (e) => {
        e.preventDefault();
        history.push("/app");
    }
    
    return (
        <Navbar className="header" >
            <NavbarBrand>
                <NavbarItem className="logo" onClick={e=>(navigateToHome(e))} >
                    <Icon className='fas fa-dove' /> 
                </NavbarItem>
                <NavbarItem className="logo-text" onClick={e=>(navigateToHome(e))}>
                    Tipoff
                </NavbarItem>
                <NavbarBurger isHidden={!currentUser.authenticated} isActive={state.isActive} onClick={(e) => onClickNav(e)} />
            </NavbarBrand>
            <NavbarMenu isHidden={!currentUser.authenticated} isActive={state.isActive} onClick={(e) => onClickNav(e)}>
                <NavbarStart></NavbarStart>
                <NavbarEnd>
                    <NavbarItem className="user-name" isHidden='touch'>
                        <Icon className='fas fa-user' />  <span >{currentUser.name}</span>
                    </NavbarItem>
                    <NavbarItem>
                        <Field isGrouped>
                            <Control>
                                <Button className="sign-out-btn" onClick={(e) => signOut(e)}>
                                    <Icon className='fas fa-power-off' /> 
                                    <span>Sign Out</span>
                                </Button>
                            </Control>
                        </Field>
                    </NavbarItem>
                </NavbarEnd>
            </NavbarMenu>
        </Navbar>
    );
    
} 
	
export default Header;