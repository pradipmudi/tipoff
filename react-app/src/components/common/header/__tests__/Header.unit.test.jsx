import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store';
import { mount} from 'enzyme';
import auth from '../../../../service/auth-svc';
import Header from '../Header';
import { useHistory } from 'react-router-dom';

const mockLogout = jest.spyOn(auth, 'logout');
mockLogout.mockResolvedValue({});

jest.mock('react-router-dom', () => ({
    __esModule: true,
    useHistory :jest.fn().mockReturnValue([])
}));

describe("Header", () => {
    describe(`logged in user view`,()=> {
        let store,initialState, wrapper;

        beforeEach (()=>{
            initialState = {
                user: { 
                    currentUser: {
                        authenticated: true,
                        name: 'Test Test'
                    }
                }
            }   
            const mockStore = configureStore();
            store = mockStore(initialState);
            wrapper = mount(<Provider store={store}><Header /></Provider>);
        });

        it("matches snapshot", () => {
            const snap = renderer.create(<Provider store={store}><Header /></Provider>).toJSON();
            expect(snap).toMatchSnapshot();
        });

        it("renders without issues", () => {
            expect(wrapper.exists()).toEqual(true);
        });

        it("renders appname correctly", () => {
            expect(wrapper.find('NavbarBrand').find('.logo-text').last().text()).toEqual('Tipoff');
        });

        it("renders username from redux store", () => {
            expect(wrapper.find('NavbarItem').find('.user-name').last().find('span').last().text()).toEqual(initialState.user.currentUser.name);
        });

        it("renders NavBar properly ", () => {
            expect(wrapper.find('Navbar').length).toEqual(1);
        });

        it("renders logo icon properly ", () => {
            expect(wrapper.find('NavbarBrand').find('Icon').hasClass('fa-dove')).toEqual(true);
        });

        it("signout button works as expected ", () => {
            const eventMock = { preventDefault: jest.fn() };
            wrapper.find('.button.sign-out-btn').props().onClick(eventMock) ;
            expect(mockLogout).toHaveBeenCalled();
        });

        it("logo icon click navigates user to homepage ", () => {
            const eventMock = { preventDefault: jest.fn() };
            wrapper.find('NavbarBrand').find('NavbarItem.logo').props().onClick(eventMock) ;
            expect(useHistory()).toContain('/app');
        });

        it("logo text click navigates user to homepage ", () => {
            const eventMock = { preventDefault: jest.fn() };
            wrapper.find('NavbarBrand').find('NavbarItem.logo-text').props().onClick(eventMock) ;
            expect(useHistory()).toContain('/app');
        });
    })

    describe(`user is not logged in view`,()=> {
        let store,initialState, wrapper;

        beforeEach (()=>{
            initialState = {
                user: { 
                    currentUser: {
                        authenticated: false,
                        name: ''
                    }
                }
            }   
            const mockStore = configureStore();
            store = mockStore(initialState);
            wrapper = mount(<Provider store={store}><Header /></Provider>);
        });

        it("matches snapshot", () => {
            const snap = renderer.create(<Provider store={store}><Header /></Provider>).toJSON();
            expect(snap).toMatchSnapshot();
        });

        it("renders without issues", () => {
            expect(wrapper.exists()).toEqual(true);
        });

        it("renders appname correctly", () => {
            expect(wrapper.find('NavbarBrand').find('.logo-text').last().text()).toEqual('Tipoff');
        });

        it("renders logo icon properly ", () => {
            expect(wrapper.find('NavbarBrand').find('Icon').hasClass('fa-dove')).toEqual(true);
        });

        it("doesn't render username ", () => {
            expect(wrapper.find('NavbarItem').find('.user-name').last().find('span').last().text()).toEqual('');
        });

        it("hides NavbarMenu ", () => {
            expect(wrapper.find('NavbarMenu').hasClass('is-hidden')).toEqual(true);
        });
       
    })
    

});