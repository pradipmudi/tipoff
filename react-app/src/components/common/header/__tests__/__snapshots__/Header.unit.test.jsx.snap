// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header logged in user view matches snapshot 1`] = `
<nav
  className="navbar header"
>
  <div
    className="navbar-brand"
  >
    <div
      className="navbar-item logo"
      onClick={[Function]}
    >
      <span
        className="icon"
      >
        <i
          aria-hidden="true"
          className="fas fa-dove"
        />
      </span>
    </div>
    <div
      className="navbar-item logo-text"
      onClick={[Function]}
    >
      Tipoff
    </div>
    <div
      className="navbar-burger"
      onClick={[Function]}
    >
      <span />
      <span />
      <span />
    </div>
  </div>
  <div
    className="navbar-menu"
    onClick={[Function]}
  >
    <div
      className="navbar-start"
    />
    <div
      className="navbar-end"
    >
      <div
        className="navbar-item is-hidden-touch user-name"
      >
        <span
          className="icon"
        >
          <i
            aria-hidden="true"
            className="fas fa-user"
          />
        </span>
          
        <span>
          Test Test
        </span>
      </div>
      <div
        className="navbar-item"
      >
        <div
          className="field is-grouped"
        >
          <div
            className="control"
          >
            <button
              className="button sign-out-btn"
              onClick={[Function]}
              type="button"
            >
              <span
                className="icon"
              >
                <i
                  aria-hidden="true"
                  className="fas fa-power-off"
                />
              </span>
              <span>
                Sign Out
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</nav>
`;

exports[`Header user is not logged in view matches snapshot 1`] = `
<nav
  className="navbar header"
>
  <div
    className="navbar-brand"
  >
    <div
      className="navbar-item logo"
      onClick={[Function]}
    >
      <span
        className="icon"
      >
        <i
          aria-hidden="true"
          className="fas fa-dove"
        />
      </span>
    </div>
    <div
      className="navbar-item logo-text"
      onClick={[Function]}
    >
      Tipoff
    </div>
    <div
      className="navbar-burger is-hidden"
      onClick={[Function]}
    >
      <span />
      <span />
      <span />
    </div>
  </div>
  <div
    className="navbar-menu is-hidden"
    onClick={[Function]}
  >
    <div
      className="navbar-start"
    />
    <div
      className="navbar-end"
    >
      <div
        className="navbar-item is-hidden-touch user-name"
      >
        <span
          className="icon"
        >
          <i
            aria-hidden="true"
            className="fas fa-user"
          />
        </span>
          
        <span>
          
        </span>
      </div>
      <div
        className="navbar-item"
      >
        <div
          className="field is-grouped"
        >
          <div
            className="control"
          >
            <button
              className="button sign-out-btn"
              onClick={[Function]}
              type="button"
            >
              <span
                className="icon"
              >
                <i
                  aria-hidden="true"
                  className="fas fa-power-off"
                />
              </span>
              <span>
                Sign Out
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</nav>
`;
